pkg/operator/hive/clustersync.go

```diff
@@ -82,22 +86,48 @@ func (r *ReconcileHiveConfig) deployClusterSync(hLog log.FieldLogger, h resource
 	if hiveconfig.Spec.MaintenanceMode != nil && *hiveconfig.Spec.MaintenanceMode {
 		hLog.Warn("maintenanceMode enabled in HiveConfig, setting hive-clustersync replicas to 0")
 		replicas := int32(0)
-		clusterSyncStatefulSet.Spec.Replicas = &replicas
+		newClusterSyncStatefulSet.Spec.Replicas = &replicas
 	} else {
 		// Default replicas to defaultClustersyncReplicas and only change if they specify in hiveconfig.
-		clusterSyncStatefulSet.Spec.Replicas = pointer.Int32Ptr(defaultClustersyncReplicas)
+		newClusterSyncStatefulSet.Spec.Replicas = pointer.Int32Ptr(defaultClustersyncReplicas)
 
 		if hiveconfig.Spec.ControllersConfig != nil {
 			// Set the number of replicas that was given in the hiveconfig
 			clusterSyncControllerConfig, found := getHiveControllerConfig(hivev1.ClustersyncControllerName, hiveconfig.Spec.ControllersConfig.Controllers)
 			if found && clusterSyncControllerConfig.Replicas != nil {
-				clusterSyncStatefulSet.Spec.Replicas = clusterSyncControllerConfig.Replicas
+				newClusterSyncStatefulSet.Spec.Replicas = clusterSyncControllerConfig.Replicas
 			}
 		}
 	}
 
-	clusterSyncStatefulSet.Namespace = hiveNSName
-	result, err := util.ApplyRuntimeObjectWithGC(h, clusterSyncStatefulSet, hiveconfig)
+	newClusterSyncStatefulSetSpecHash, err := controllerutils.CalculateStatefulSetSpecHash(newClusterSyncStatefulSet)
+	if err != nil {
+		hLog.WithError(err).Error("error calculating new statefulset hash")
+		return err
+	}
+
+	if newClusterSyncStatefulSet.Annotations == nil {
+		newClusterSyncStatefulSet.Annotations = make(map[string]string, 1)
+	}
+	newClusterSyncStatefulSet.Annotations[hiveClusterSyncStatefulSetSpecHashAnnotation] = newClusterSyncStatefulSetSpecHash
+
+	existingClusterSyncStatefulSet := &appsv1.StatefulSet{}
+	existingClusterSyncStatefulSetNamespacedName := apitypes.NamespacedName{Name: newClusterSyncStatefulSet.Name, Namespace: newClusterSyncStatefulSet.Namespace}
+	err = r.Get(context.TODO(), existingClusterSyncStatefulSetNamespacedName, existingClusterSyncStatefulSet)
+	if err == nil && hasStatefulSetSpecChanged(existingClusterSyncStatefulSet, newClusterSyncStatefulSet, hLog) {
+		// The statefulset spec.selector changed. Trying to apply this new statefulset will error with:
+		//     invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden"
+		// The only fix is to delete the statefulset and have the apply below recreate it.
+		hLog.Info("deleting the existing clustersync statefulset because spec has changed")
+		err := r.Delete(context.TODO(), existingClusterSyncStatefulSet)
```

Comment:
what's the default forf delete in terms of waiting for deletion propagation, are we waiting for all pods and the object to be deleted before moving ahead.

what happens if we delete, the deletion timestamp is set and object is not yet deleted and we try to create a new one immediately after.

Reply:
Would a simple solution to this possible problem be to setup a watch for our StatefulSet that triggers a reconcile, and then in this logic we return immediately after issuing the Delete, and probably add a check for DeletionTimestamp already being set, in which case we also return right away. Basically just keep reconciling if we see the spec changed and we're not fully deleted, once it's gone we know to recreate it.

Reply:
Bearing in mind that kind of watch is a long standing goal of hive-operator we haven't tackled yet. We'd love for it to react to changes to the resources it creates so it could repair them.

Reply:
@dgoodwin are you proposing those changes for this PR or a future PR when we switch to watching resources we create?

I'm fine doing the work on this, but we should likely create a card as it seems like an arch change, not a bug.

Reply:
I'm proposing just the addition of a watch for this StatefulSet, and the return early logic proposed above, I did not consider this an architectural change, rather an improvement to this PR to prevent a potential bug. We already watch more things than I realized in hive_controller.go, but StatefulSet is not one of them.

Reply:
Ah, ok. From your description, I thought it was a bigger change.

Reply:
@dgoodwin I believe I've implemented your proposed fix for this one. PTAL.